### DISCUSSION TITLE
Grid Search for DeepAR

### DISCUSSION THREAD

**Discussion by parimuns** [awslabs/gluonts#1417] [General]
I am trying to tune the hyperparameters of DeepAr with Grid Search method .But when I call the method ,it gives me error 
'TypeError: estimator should be an estimator implementing 'fit' method'.
I am first time user of grid search ,so don't know exactly where am I doing the mistake.
I am attaching the snippet of my code as below

```
%matplotlib inline
!pip install mxnet
!pip install gluonts
import mxnet as mx
import gluonts
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import json
import os
from itertools import islice
from pathlib import Path
from sklearn.model_selection import GridSearchCV
from gluonts.dataset.common import FileDataset, TrainDatasets, load_datasets
from gluonts.dataset.field_names import FieldName
[f"FieldName.{k} = '{v}'" for k, v in FieldName.__dict__.items() if not k.startswith('_')]
train_data=pd.read_csv(r'/content/drive/MyDrive/Colab Notebooks/ISO_NE_train.csv')
test_data =pd.read_csv(r'/content/drive/MyDrive/Colab Notebooks/ISO_NE_test.csv')
from gluonts.model.deepar import DeepAREstimator
from gluonts.trainer import Trainer
from gluonts.model.deepar import DeepAREstimator
from gluonts.trainer import Trainer
from gluonts.distribution import NegativeBinomialOutput
estimator=DeepAREstimator(freq='H',
        prediction_length=12,
        context_length=24,
        distr_output = NegativeBinomialOutput(),
        cardinality=[8],
        trainer= Trainer(
            epochs=100, num_batches_per_epoch=16, hybridize=False,learning_rate=0.05,learning_rate_decay_factor=0.9),
        num_layers= 2,
        num_cells= 40,
        cell_type= "lstm",
        num_parallel_samples = 100,
        dropout_rate= 0.1,
        scaling=True,
        use_feat_dynamic_real = False,
        use_feat_static_cat = True,
    )
num_batches_per_epoch=[8,16,32,64,128,256]
epochs=[50,100,150]

param_grid=dict(batch_size=num_batches_per_epoch, epochs=epochs)
mygrid=GridSearchCV(estimator=estimator,param_grid=param_grid,n_jobs=-1,cv=3)
grid_result=mygrid.fit(train_data,test_data)
```

**Comment by kaijennissen** [comment]
Hi. The `GridSearchCV` class from `scikit-learn` is intended to be used with a scikit-like estimator, not with the DeepAREstimator or any other model from gluonts. For hyperparamter tuning with gluonts you have to write your own loop.

**Reply by parimuns** [reply]
Hello @kaijennissen.Thanks a lot for answering. Can you please guide me, how can we write it? I am new to python programming, so don't have much expertise. Thanks a lot for your help.

**Reply by nils-fl** [reply]
You can basically use any package where you define an objective and a procedure to walk though your parameter space.
Do it on your own or have a look at Optuna.
You define an objective - would be to train and get your MSE or any other metric - and your parameter space and then you run a random search (sample from your parameter space) or TPE to minimize (or maximize) your metric/objective.
